Generate extra imports in local compilation mode for bundling purposes

pmvald opened the pull request:
The particular way that bundling is done in g3 requires that prod and dev sources have same or compatible import statements. Since the dev sources are compiled in local mode, then this requires local compilation to generate extra imports which imply the ones generated in full compilation mode due to resolving component's dependencies. 

A new g3 specific option `generateExtraImportsInLocalMode` is added, which when is set and the compilation mode is local the compiler will generate extra imports. 

The extra imports are generated in a way that the the prod sources can be bundled based on these extra imports plus user written imports without issue. This requires to perform static analysis as much as possible in local mode to resolve the component dependencies that are within the same target. For this to happen it is assumed that we are using local compilation mode per target, and not when per single (which is the case in g3). For external deps, we spread all external imports that could be candidate for exposing an external dependency to all the all the files in the compilation unit.

TGP when local compilation mode is enabled for dev sources plus this change: [link](https://fusion2.corp.google.com/presubmit/587124714/OCL:587124714:BASE:599408150:1705564041612:3f5dc3b9/targets)


## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

pmvald committed:
refactor(compiler-cli): add API for extra import tracker

The tracker is responsible for registering the extra imports during the analysis and resolve compiler phases, and later to be used by the transformer to get a list of extra imports to be generated for each source file.

This commit only contains the API, and the actual implementation for each method will be done in subsequent commits where an application of that method is available and so tests can be written for the implementation.

pmvald committed:
refactor(compiler-cli): incorporate `LocalCompilationExtraImportsTracker` in the compilation workflow

This commit includes a skeleton of how the tool `LocalCompilationExtraImportsTracker` is used in the overall compilation workflow end-to-end.

First of all, a new option `generateExtraImportsInLocalMode` is added, whose presence will make `LocalCompilationExtraImportsTracker` part of the compilation process. When this option is set an instance of `LocalCompilationExtraImportsTracker` is created within the NgCompiler. Then it is passed to the Ivy transformer and plumbed all the way down and the extra imports registered in it are added to the `ImportManager` instances before the imports are added from `ImportManager` to the generated file. This required adding a new method `generateSideEffectImport` to the `ImportManager`, which is an empty method and will be implemented in the subsequent commits.

This commit expected to make no change in the compilation behavior as the methods are not implemented yet.

pmvald committed:
refactor(compiler-cli): ability to generate side effect imports

As the first step,  the import manager's `generateSideEffectImport` method is implemented to enable it to store info for side effect imports. Next, the helper `addImports` is modified to be able to generate correct statement for side effect imports.

These changes will be tested in the subsequent commits when these tools are used to generate an actual extra import for the generated file.

pmvald committed:
feat(compiler-cli): generate global imports in local compilation mode

With option `generateExtraImportsInLocalMode` set in local compilation mode, the compiler generates extra side effect imports using this rule: any external module from which an identifier is imported into an NgModule will be added as side effect import to every file in the compilation unit. To illustrate this better assume the compilation unit has source files `a.ts` and `b.ts`, and:

```
// a.ts
import {SomeExternalStuff} from 'path/to/some_where';
import {SomeExternalStuff2} from 'path/to/some_where2';

...

@NgModule({imports: [SomeExternalStuff]})

```

then the extra import `import "path/to/some_where"` will be added to both `a.js` and `b.js`. Note that this is not the case for `import "path/to/some_where2"` though, since the symbol `SomeExternalStuff2` is not imported into any NgModule.

The math behind this mechanism is, in local compilation mode we cannot resolve component external dependencies fully. For example if a component in `a.ts` uses an external component defined in an external file `some_external_comp.ts` then we can generate the import to this file in `a.js`. Instead, we want to generate an import to a file that "gurantees" that `a.js` is placed after `some_external_comp.js` in the bundle. Now since the component in  `some_external_comp.ts` is used in `a.ts`, then there must be a chain of imports starting from the NgModule that declares the component in `a.ts` to the component in `some_external_comp.ts`. This chain means some file in the same compilation unit as `a.ts` should import some external NgModule which includes `some_external_comp.ts` in its transitive closure and import it to some NgModule. So by adding this import to `a.js` we ensure that the bundling will have the right order.

pmvald committed:
refactor(bazel): make option generateExtraImportsInLocalMode configurable

This is to allow testing this option using bazel

pmvald committed:
refactor(compiler-cli): run resolve phase fully for components in local mode when `generateExtraImportsInLocalMode` is set

In this commit the resolve method for components is run fully when the option `generateExtraImportsInLocalMode` is set. This is because we need local component depedencies in order to generate extra imports causing by them. This requires cutting some resolve phase logics that are unnecessary in local mode, such as diagnostics.

pmvald committed:
feat(compiler-cli): generate extra imports for component local dependencies in local mode

With option `generateExtraImportsInLocalMode` set, in local mode the compiler generates extra imports for each component local dependencies. Here local dependencies means all component's dependencies within the same compilation unit.

To achieve this, the compiler performs a "local version" of its regular static analysis to find each component's deps, and these deps are used to generate extra side effect imports.

pmvald committed:
fix(compiler-cli): generating extra imports in local compilation mode when cycle is introduced

At the moment the extra import generation in local compilation mode fails if these extra imports produce a cycle. To handle this, the cycle handling strategy is updated for local compilation, and following the behaviour in the full compilation mode, the compiler does not generate extra import if it leads to cycle and instead leave things to the runtime.

devversion reviewed (commented):
Leaving some comments here if we end up going with this approach. One thing I want us to consider is using the existing "aliasing" logic that is also used for full compilation. This would feel more natural and maybe avoid lots of extra code. Have you spent some time thinking about this? I feel like this would be better if we can get it to work somehow. We'd just need to resolve the scopes as much as possible (which is what are kind of doing with this PR already) 

https://github.com/angular/angular/blob/main/packages/compiler-cli/src/ngtsc/imports/src/alias.ts#L60

https://github.com/angular/angular/blob/main/packages/compiler-cli/src/ngtsc/scope/src/local.ts#L540

pmvald commented on packages/compiler-cli/src/ngtsc/annotations/ng_module/src/handler.ts:
```diff
@@ -267,17 +270,33 @@ export class NgModuleDecoratorHandler implements
-      exportRefs = this.resolveTypeList(rawExports, exportsMeta, name, 'exports', 0).references;
+      exportRefs = this.resolveTypeList(
+                           rawExports, exportsMeta, name, 'exports', 0,
+                           this.compilationMode === CompilationMode.LOCAL)
```
No. But if we set this arg to false then `resolveTypeList` throws exception the moment one value resolved to DynamicValue. The current behaviour of `resolveTypeList` is:

1) last arg is false: don't allow any DynamicValue
2) last arg is true: allow DynamicValues and also collect them in case needed

ATM, feature (2) is only needed for imports. But this way we can use resolveTypeList for all three fields imports, exports and declarations.

We can of course teach resolveTypeList not to collect DynamicValues, and it may save some perf, but might not be much.

devversion replied:
is this something we want to support though? or can we just avoid this if completely if we are in local mode, like it was before this PR?

pmvald replied:
You mean support for collecting dynamic values? Yes, they are needed for `imports`. Dynamic value on NgModule imports are candidates for global extra imports. Because if an NgModule imports an external NgModule, chances are that external NgModule contributes some component to the current target. So we add its import to all the files.

devversion replied:
But in this case here, this is just collecting `exportRefs`? Without digging much into it right now, it felt like `exportRefs` is not needed in local compilation as it's not used? (or why did we have the check in place before?)

pmvald replied:
Yes, because we use the same signature it collects export dynamic values, but we have no use for them. We can add a new arg or option to `resolveTypeList` to control this behaviour as to whether to collect or not. We need collection only for import. For all others we just want ignoring dynamic values and not throw. Is this what you suggesting? However, collecting dynamic values does not seem to be much load bearing in terms of performance or code complexity and somehow come as part of the code path itself.

devversion replied:
oh sorry, I was wondering why this code path is executed at all for local compilation? i.e. should simply revert and not call resolveTypeList at all here (like it was before)

devversion replied:
In either case though, I'm good with this. Just trying to understand the reasoning better

pmvald replied:
Ah I see what you mean. Yes, we need importRefs, exportRefs and declarationRefs in order to compute the NgModule scope. Using that scope we calculate the so-called `eagerDeclarations` for each component which are depedencies of that component. In local mode the importRef, exportRef, etc are a subset of actual ones which contain only the local references. As a result NgModule scope and component eagerDeclarations also become local subsets. 

Then we use these (local) component dependencies to compute the extra imports for each component corresponding to its local deps [ref](https://github.com/angular/angular/pull/53543/files#diff-d31c7054eab6f7fd566cfc0dc6df292a9768ccbcb76475fa2da9bb66797df120R867)

pmvald commented:
> Leaving some comments here if we end up going with this approach. One thing I want us to consider is using the existing "aliasing" logic that is also used for full compilation. This would feel more natural and maybe avoid lots of extra code. Have you spent some time thinking about this? I feel like this would be better if we can get it to work somehow. We'd just need to resolve the scopes as much as possible (which is what are kind of doing with this PR already)
> 
> https://github.com/angular/angular/blob/main/packages/compiler-cli/src/ngtsc/imports/src/alias.ts#L60
> 
> https://github.com/angular/angular/blob/main/packages/compiler-cli/src/ngtsc/scope/src/local.ts#L540

Is this the API that generated re-exports ? We want to turn this off ideally to allow interop between local and global. I already sent you a doc for that.

devversion commented:
This would be even better, either re-using that logic, or disabling it completely sounds even better.

pmvald commented on packages/compiler-cli/src/ngtsc/transform/src/compilation.ts:
```diff
@@ -596,28 +601,29 @@ export class TraitCompiler implements ProgramTypeCheckAdapter {
         }
-        // `trait.resolution` is non-null asserted below because TypeScript does not recognize that
-        // `Readonly<unknown>` is nullable (as `unknown` itself is nullable) due to the way that
-        // `Readonly` works.
```
The first one is not needed, since I changed `trait.analysis!` -> `trait.analysis` and no problem. Second one seems still needed. So added it back.

devversion commented on packages/compiler-cli/src/ngtsc/transform/src/compilation.ts:
```diff
@@ -596,28 +601,29 @@ export class TraitCompiler implements ProgramTypeCheckAdapter {
     for (const trait of record.traits) {
       let compileRes: CompileResult|CompileResult[];
-      if (this.compilationMode === CompilationMode.LOCAL) {
+      if (!this.shouldRunResolvePhase) {
```
I was imaging we'd still have a safety check for readability/clarity

pmvald replied:
Ok, I tried. I becomes like this.

alxhub commented on packages/compiler-cli/src/main.ts:
```diff
@@ -29,6 +29,8 @@ export function main(
     return reportErrorsAndExit(result.firstCompileResult, options, consoleError);
   }
 
+  // console.warn(">>>> rootNames", rootNames);
```
💣

alxhub commented on packages/compiler-cli/src/ngtsc/transform/src/compilation.ts:
```diff
@@ -95,6 +95,20 @@ export class TraitCompiler implements ProgramTypeCheckAdapter {
+  // In local compilation mode the resolve phase is needed only when we want to generate extra
+  // imports (see the comment for {@link shouldRunResolvePhase}). To run the resolve phase, we will
+  // need to run the register phase beforehand.
+  private readonly shouldRunRegisterPhase =
```
What do you think about always running all the phases, and moving the logic of handling local mode into the `resolve` functions themselves (sort of like we did in `PipeDecoratorHandler`, etc)? That keeps the `TraitCompiler` from needing any special knowledge of when these phases are needed.

pmvald replied:
I guess Andrew is going to make a separate PR for this. So I wait for that and then remove this commit altogether.

AndrewKushnir replied:
FYI, here is the PR with the change that enables `resolve` phase for local compilation: https://github.com/angular/angular/pull/53901.

pmvald replied:
Done!

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/ng_module/src/handler.ts:
```diff
@@ -832,11 +857,24 @@ export class NgModuleDecoratorHandler implements
       expr: ts.Node, resolvedList: ResolvedValue, className: string, arrayName: string,
-      absoluteIndex: number):
-      {references: Reference<ClassDeclaration>[], hasModuleWithProviders: boolean} {
+      absoluteIndex: number, allowUnresolvedReferences = false): {
```
In the compiler we generally avoid optional parameters. It's fine to make this a required parameter and update existing usage sites.

pmvald committed:
refactor(compiler-cli): run register phase in local compilation mode

When option `generateExtraImportsInLocalMode` is set, we need to compute component local depednecies in order to generate extra imports related to them. At the same time running the register phase in general is harmless in local compilation. So we run it anyway.

devversion commented:
@pmvald do we actually still need this PR if we disable extra imports in global compilation?

pmvald commented:
> @pmvald do we actually still need this PR if we disable extra imports in global compilation?

Which disabling extra imports? You mean #53937? No, they are two different things. This change is about the extra imports generated by statically resolving component deps (and thus adding extra imports to import these deps). These extra imports (or a super set of them) should be also present in local mode just to satisfy our bundler. But #53937 is about extra reexports generated apparantly to expose each NgModule's scope members in teh same file as the NgModule is defined. These extra reexports are harmless per sei. The only problem is they don't allow a globally compiled target to have dependencies on locally compiled target (since the parent target will want to import from these extra reexports in the child which do not exist in local mode). So #53937 is merely for the interop sake. But this PR is essential to bundle the prod sources correctly. It is the main blocker for g3.

devversion commented:
Oh makes sense. Thanks for clarifying again. I think I messed up these two things, thinking this would not be needed. Will review again tomorrow morning

devversion commented on packages/compiler-cli/src/ngtsc/annotations/ng_module/src/handler.ts:
```diff
@@ -267,17 +269,31 @@ export class NgModuleDecoratorHandler implements
+          rawImports, importsMeta, name, 'imports', 0,
+          this.compilationMode === CompilationMode.LOCAL);
+
+      if (this.extraImportsTracker !== null) {
```
Add a comment on why this is done. Ideally even a link to your design doc + a comment capturing why this is useful / how it solves the issue

devversion commented on packages/compiler-cli/src/ngtsc/annotations/ng_module/src/handler.ts:
```diff
@@ -267,17 +269,31 @@ export class NgModuleDecoratorHandler implements
+
+      if (this.extraImportsTracker !== null) {
+        for (const d of result.dynamicValues) {
+          this.extraImportsTracker.addGlobalImportFromIdentifier(d.node);
```
Why are we doing an import in every file? Is this just for simplicity? Technically I'd expect it to be sufficient only adding these imports to files depending on this module.

Please add a comment to the documentation of `addGlobalImportFromIdentifier`, and also re-consider if we can actually optimize this (for our minds)

devversion reviewed (approved):
LGTM, minor comments/suggestions. Clever solution!

pmvald commented on packages/compiler-cli/src/ngtsc/annotations/component/src/handler.ts:
```diff
@@ -943,13 +956,21 @@ export class ComponentDecoratorHandler implements
-            DeclarationListEmitMode.Closure :
-            DeclarationListEmitMode.Direct;
+
+        // Register extra local imports.
```
Done!

Renamed to LocalCompilationExtraImportTracker

devversion reviewed (approved):
LGTM.

Caretaker note: No need for public API approval since this is a compiler option and the compiler is not considered as part of the public API (and it's a new option)

jessicajaniuk commented:
This PR was merged into the repository by commit 64fa5715c696101fba0b4f8623eaec0eadc5b159.
